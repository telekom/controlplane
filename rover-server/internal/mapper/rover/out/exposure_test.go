// Copyright 2025 Deutsche Telekom IT GmbH
//
// SPDX-License-Identifier: Apache-2.0

package out

import (
	"github.com/gkampitakis/go-snaps/snaps"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"

	roverv1 "github.com/telekom/controlplane/rover/api/v1"

	"github.com/telekom/controlplane/rover-server/internal/api"
)

var _ = Describe("Exposure Mapper", func() {
	Context("mapApiExposure", func() {
		It("must map ApiExposure correctly", func() {
			input := &apiExposure

			output := mapApiExposure(input)

			Expect(output).ToNot(BeNil())
			snaps.MatchSnapshot(GinkgoT(), output)
		})

		It("must handle empty ApiExposure input", func() {
			input := &roverv1.ApiExposure{}

			output := mapApiExposure(input)

			Expect(output).ToNot(BeNil())
			snaps.MatchSnapshot(GinkgoT(), output)
		})
	})

	Context("mapExposure", func() {
		It("must map an ApiExposure correctly", func() {
			input := GetApiExposure(&apiExposure)
			output := &api.Exposure{}

			err := mapExposure(&input, output)

			Expect(err).To(BeNil())
			snaps.MatchSnapshot(GinkgoT(), output)
		})

		It("must map an EventExposure correctly", func() {
			input := GetEventExposure(&eventExposure)
			output := &api.Exposure{}

			err := mapExposure(&input, output)

			Expect(err).To(BeNil())
			snaps.MatchSnapshot(GinkgoT(), output)
		})

		It("must return an error for unknown exposure type", func() {
			input := &roverv1.Exposure{}
			output := &api.Exposure{}

			err := mapExposure(input, output)

			Expect(err).ToNot(BeNil())
			Expect(err.Error()).To(ContainSubstring("unknown exposure type"))
			snaps.MatchSnapshot(GinkgoT(), output)
		})
	})

	Context("toApiVisibility", func() {
		It("must map WORLD visibility correctly", func() {
			input := roverv1.VisibilityWorld

			output := toApiVisibility(input)

			Expect(output).To(Equal(api.WORLD))
		})

		It("must map ZONE visibility correctly", func() {
			input := roverv1.VisibilityZone

			output := toApiVisibility(input)

			Expect(output).To(Equal(api.ZONE))
		})

		It("must map ENTERPRISE visibility correctly", func() {
			input := roverv1.VisibilityEnterprise

			output := toApiVisibility(input)

			Expect(output).To(Equal(api.ENTERPRISE))
		})

		It("must map unknown visibility", func() {
			input := roverv1.Visibility("Unknown")

			output := toApiVisibility(input)

			Expect(output).To(Equal(api.Visibility("UNKNOWN")))
		})

		Context("toApiApprovalStrategy", func() {
			It("must map AUTO approval strategy correctly", func() {
				input := roverv1.ApprovalStrategyAuto

				output := toApiApprovalStrategy(input)

				Expect(output).To(Equal(api.AUTO))
			})

			It("must map MANUAL approval strategy correctly", func() {
				input := roverv1.ApprovalStrategySimple

				output := toApiApprovalStrategy(input)

				Expect(output).To(Equal(api.SIMPLE))
			})

			It("must map FOUREYES approval strategy correctly", func() {
				input := roverv1.ApprovalStrategyFourEyes

				output := toApiApprovalStrategy(input)

				Expect(output).To(Equal(api.FOUREYES))
			})

			It("must map unknown approval strategy", func() {
				input := roverv1.ApprovalStrategy("Unknown")

				output := toApiApprovalStrategy(input)

				Expect(output).To(Equal(api.ApprovalStrategy("UNKNOWN")))
			})
		})
	})
})
