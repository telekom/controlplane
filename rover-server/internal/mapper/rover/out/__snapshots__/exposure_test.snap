
[Exposure Mapper mapApiExposure must map ApiExposure correctly - 1]
api.ApiExposure{
    Approval:      "SIMPLE",
    BasePath:      "/eni/distr/v1",
    Failover:      api.Failover{},
    LoadBalancing: api.LoadBalancing{},
    RateLimit:     api.RateLimitContainer{},
    RemoveHeaders: nil,
    Security:      api.Security{},
    TrustedTeams:  nil,
    Type:          "",
    Upstream:      "https://httpbin.org/anything",
    Visibility:    "WORLD",
}
---

[Exposure Mapper mapApiExposure must handle empty ApiExposure input - 1]
api.ApiExposure{
    Approval:      "",
    BasePath:      "",
    Failover:      api.Failover{},
    LoadBalancing: api.LoadBalancing{
        Health:  api.Health{},
        Servers: {
        },
    },
    RateLimit:     api.RateLimitContainer{},
    RemoveHeaders: nil,
    Security:      api.Security{},
    TrustedTeams:  nil,
    Type:          "",
    Upstream:      "",
    Visibility:    "",
}
---

[Exposure Mapper mapExposure must map an ApiExposure correctly - 1]
&api.Exposure{
    union: {0x7b, 0x22, 0x61, 0x70, 0x70, 0x72, 0x6f, 0x76, 0x61, 0x6c, 0x22, 0x3a, 0x22, 0x53, 0x49, 0x4d, 0x50, 0x4c, 0x45, 0x22, 0x2c, 0x22, 0x62, 0x61, 0x73, 0x65, 0x50, 0x61, 0x74, 0x68, 0x22, 0x3a, 0x22, 0x2f, 0x65, 0x6e, 0x69, 0x2f, 0x64, 0x69, 0x73, 0x74, 0x72, 0x2f, 0x76, 0x31, 0x22, 0x2c, 0x22, 0x74, 0x79, 0x70, 0x65, 0x22, 0x3a, 0x22, 0x61, 0x70, 0x69, 0x22, 0x2c, 0x22, 0x75, 0x70, 0x73, 0x74, 0x72, 0x65, 0x61, 0x6d, 0x22, 0x3a, 0x22, 0x68, 0x74, 0x74, 0x70, 0x73, 0x3a, 0x2f, 0x2f, 0x68, 0x74, 0x74, 0x70, 0x62, 0x69, 0x6e, 0x2e, 0x6f, 0x72, 0x67, 0x2f, 0x61, 0x6e, 0x79, 0x74, 0x68, 0x69, 0x6e, 0x67, 0x22, 0x2c, 0x22, 0x76, 0x69, 0x73, 0x69, 0x62, 0x69, 0x6c, 0x69, 0x74, 0x79, 0x22, 0x3a, 0x22, 0x57, 0x4f, 0x52, 0x4c, 0x44, 0x22, 0x7d},
}
---

[Exposure Mapper mapExposure must map an EventExposure correctly - 1]
&api.Exposure{
    union: {0x7b, 0x22, 0x61, 0x70, 0x70, 0x72, 0x6f, 0x76, 0x61, 0x6c, 0x22, 0x3a, 0x22, 0x22, 0x2c, 0x22, 0x65, 0x76, 0x65, 0x6e, 0x74, 0x54, 0x79, 0x70, 0x65, 0x22, 0x3a, 0x22, 0x74, 0x61, 0x72, 0x64, 0x69, 0x73, 0x2e, 0x68, 0x6f, 0x72, 0x69, 0x7a, 0x6f, 0x6e, 0x2e, 0x64, 0x65, 0x6d, 0x6f, 0x2e, 0x63, 0x65, 0x74, 0x75, 0x73, 0x2e, 0x76, 0x31, 0x22, 0x2c, 0x22, 0x74, 0x79, 0x70, 0x65, 0x22, 0x3a, 0x22, 0x65, 0x76, 0x65, 0x6e, 0x74, 0x22, 0x2c, 0x22, 0x76, 0x69, 0x73, 0x69, 0x62, 0x69, 0x6c, 0x69, 0x74, 0x79, 0x22, 0x3a, 0x22, 0x22, 0x7d},
}
---

[Exposure Mapper mapExposure must return an error for unknown exposure type - 1]
&api.Exposure{}
---
